README
Fonty Python
Copyright (C) 2006,2007,2008,2009,2016 Donn.C.Ingle
donn.ingle@gmail.com

Fonty Python comes with ABSOLUTELY NO WARRANTY; for details
see the COPYING file.  This is free software, and you are welcome
to redistribute it under certain conditions; see the COPYING file 
for details.

NB: Remove the old version BEFORE you install FP.
==============
You can find the path by:
locate fontypythonmodules

If that does not work, then you must update your locate database:
As root run: locate -u
Go make coffee.Then run the locate command above again.

This will give you a directory something like:
/usr/lib/python2.5/site-packages/fontypythonmodules

Use your root account to remove that folder:
sudo rm -fr /usr/lib/python2.5/site-packages/fontypythonmodules

(Be *very* careful what you type!)
You should be good to go now.

How to install
==============
Please see the end of this file if your install fails, there are
certain files that FP relies upon.

I assume you have extracted the "tarball" already.
Change to the directory it creates.

As the root user enter this command:
python setup.py install

If you are on Ubuntu/Kubuntu/Xubuntu/Debian etc:
sudo python setup.py install

This will create a program called 'fontypython'

Alternative installation
========================
Move the entire extracted folder to a folder that
is in your PATH, like ~/bin
Make a link to the fontypython script like this: (for e.g.)
tar -s ~/bin/fontypython-x.x.x/fontypython .
Make it executable:
chmod u+x fontypython

Now you should be good to go.
If it does not work, try closing your console and
starting it again.

Running the program
===================
To get started:
fontypython -h

For more info:
fontypython -e

Please read the help ( -h or Help menu in app ).

Troubleshooting the installation
================================
Fonty Python depends on several other libraries:
( These are minimum version numbers )
1. python2.4 to python2.6 (Not python3)
2. python-dev (odd one this. See notes below)
3. python-imaging
4. python-wxgtk2.6 or greater.

Items 1 and 3 are usually pre-installed on Gnu\Linux distros.
Item 2 is a new twist. I am not sure why it's required for FP
(yes - I don't know everything :)) but some users have had
problems and installing python2.4-dev fixed them.
(replace 2.4 with the actual version of python you are using.)
On *buntu:
sudo apt-get install python2.4-dev

If you cannot install item 2 - try the alternative 
installation instructions above.

Item 4 may be installed, or it may be in your distro's
main repository.To install item 4, try the following:
On *buntu/Debian:
apt-cache search python-wxgtk
Find the latest version in the list, then:
sudo apt-get install python-wxgtkX.Y
(replacing X.Y with the results of your search)

On distro X:
*Please send me your experiences and I'll fill this in*

After this has installed, try to run fontypython again.

If you still do not come right, then I advise you
to go to the following web sites for help directly
from the horse's mouth:
http://wxpython.org
http://www.pythonware.com/products/pil/index.htm


LOCALIZATION TIPS
=================
Make sure your LANG variable is set properly.
To run under a locale, do something like this:
LANG=en_ZA.utf8
./fontypython

You can find your locales by:
locale -a

If localization is not working it could be that there is no 
translation for your language yet, it can also be a 
problem with missing packages in your distro.

This is what I installed on my system (Kubuntu 7.10 as of 
December 2007) while I was developing:
 
language-support-fr
language-support-en
language-pack-gnome-fr (*)
(*) This one is VERY important, it has many stock translations for GTK. Without it
you likely not see buttons and widgets in the right language.

Substitute your language code as you please.

TRANSLATION TIPS
================
If you want to help translate, please contact us via the 
fontypython list: fontypython@googlegroups.com

I have included the pot files that I have. Look at fontypythonmodules/pofiles/README
(Translation stuff is messy. Headers are all wrong and versions all over the place.)



CHANGELOG
Changelist
==========
Latest entries at the bottom.

Jan 4 2008 : FP version 0.3.1
* Fixed img height/width bugs in help file. It scrolls smoothly now.
* Fixed startup bug when last folder viewed no longer has any fonts.
* Fixed some text in strings.
* Improved the setup.py system to use i18n, strings and fpversion modules.
* Added Italian translation (Thanks to Pietro Battiston)

Jan 10 2008 : FP version 0.3.3
* Fixed setup.py bug - should install properly now.
* Tweaked some gui things - like spacing.
* Added a clear button for the filter.
* Major overhaul of the Type1 code in fontcontrol.

Jan 18 2008 : FP version 0.3.4
* Did a lot of locale/unicode work to support LANG=C
* Added a tool to 'check' fonts - marks segfaulters
* Added long command-line options
* Repairs to help file
* Many bugs fixed, new ones created :)

Jan 20 2008 : FP version 0.3.5
* Added a .desktop file and an icon - changes to setup.py

Jan 22 2008 : FP version 0.3.6
* Fixed another font bug (0 width).
* Bad wxImage badfont catch - draws a yellow block like a segfont.
* Made the Check Font dialog work a lot better. feedback & better layout.
* Updated the po/mo files, but no new translating has been done.

11 June 2009
1. Changed many button events to wx.EVT_BUTTON for better behaviour )
2. Fixed Bug 2 (wxgui) -- The Tab->Folder->Last Directory->Show Fonts bug.
3. Changed cli, fontcontrol, strings and wxgui to allow recursive folders. 

12 June 2009
1. Adding a Search Assistant panel. It's full of junk from the wxDemo at the moment.
2. Added/Edited ROADMAP after reading email from Michael.
3. Removed Select All button. RIP bad idea :)
4. Added Selection menu and it's sub-items. Wired-them up to the remains of the Select Button code.
5. Fixed the PogChooser issue betwixt source and target. 
6. Split-off most of wxgui.py classes into their own gui_xxx.py modules. 
   Trying to make it simpler. We now have gui_Left, gui_Middle and gui_Right for the main sections.

13 June 2009
1. Continued the refactoring. Moved functions to where they made more sense. Cut down on chains of calls.
2. Got rid of shadow functions. At last!
3. Put refs to locale stuff into the new modules, BUT HAVE NOT TESTED i18n yet. HELP!
4. Improved the pubsub a tiny little bit. Also removed the return value -- it was not being used.
5. Fixed strings GPL message. It loads it from the COPYING file.
6. Hacking a search bar together from a combo box and some spare bits :)

14 June
1. Got the search combo box working a little. History is added etc.

16 June
1. Fixed the bug where the first fitmap in the fontview would draw at an odd y value.
2. Fixed the persistent InfoFontItem issue. Now it goes away when it should.

18 June
1. Many visual fixes. Re-factored the drawing code in Fitmap.
2. Many little fixes like toggle selection menu and multi-selection on/off in the lists.

19 June
1. Gave up on the fancy search dream. FontTools and Panose confuse me horribly. 
   Michael can't help dev at the moment, so it's all back to square one.
   I remarked-out the assistant panel.
2. I got the left and right arrow keys to step through the list - like the buttons do.
3. I (think I) fixed the ESC key to work from anywhere.

20 June
1. Cleaning-up files etc. Lot's of little zipzap commits. Yesterday I merged the branch
   back into the trunk.
2. Caught an encode and decode problem in Folder.__init__. Had to jump through the usual
   unicode hoops to fix it. Hope it is fixed now...
3. Added the licence header to a bunch of files it was missing from.
4. Small fixes to README regarding i18n.
5. Fixing some unicode errors and error dialog boxes.
6. Trying to repair the "lost focus" bug when the main button is clicked -- so far no dice.

22 June
1. Rigged the main button so that focus is not lost on click. It's a hack, but what the heck.

26 June
1. Main button focus issue has a hack-fix
2. Delete multiple pogs at once from the gui.
3. Fixed the help text. Changed the app screenshot.
4. Moved the Purge function out to the File menu. It now applies to the VIEW POG selected.
26 June 
1. Added a test to prevent fonts re-drawing when a new target Pog is clicked.
2. Added help about the Purge menu item.
26 June 
1. Reversed decision to not redraw the target Pog. There are too many subtleties involved for the hammer-and-nail approach I took and I just don't have time to be subtle now. So, it will redraw when you choose another target Pog. Tough.

27 June
1. Changed left/right arrow to pageup/pagedown. Michael Hoeft warned of the damage to editing the search term.
2. Improving the splash image.
27 June
1. Fixed ref to page up/down keys in help.

28 June
1. Upon reflection, the arrow keys are easier than page up/down. So, I added CTRL as a modifier. Now it's ctrl+left/right to step through the pages. (Help updated too.)

29 June
1. Repaired the .desktop file to comply with HIG.

30 June
1. Changed selection menu item toggling system. Recently read a little Gnome HIG :)

1 July
1. Changed logo in help, about and splash.
2. Altered setup.py to remove MANIFEST file if it's there. Unless it's removed, setup.py sdist will potentially miss new files.

6 July 2009
1. Fixed a problem with setup.py that was making ~/.fontypython folder (and as root too). That folder now only gets made when the app is run for the first time, not during setup. Thanks to Pietro for catching this one.

7 July
1. Pietro sent us an Italian .po file. I did "make mos" and I hope that was right.
2. Removed some _() stuff from CORNER CASE error messages.
3. Redacted pofiles/README file somewhat.

24 July
1. A bug was caught by Michael Hoeft - the combo box (for search field) requires wxPython 2.8 -- So we had to shift a few things around. After a false start, I have created version 0.4.1 and am about to upload to the ftp site.

15 Sept 2009
1. Fixed a bug where fonts that are already installed would refuse to install again. It ignores OSError 17 now.

fontypythonmodules/strings.py
# This Python file uses the following encoding: utf-8
##	Fonty Python Copyright (C) 2006, 2007, 2008, 2009 Donn.C.Ingle
##	Contact: donn.ingle@gmail.com - I hope this email lasts.
##
##	This file is part of Fonty Python.
##	Fonty Python is free software: you can redistribute it and/or modify
##	it under the terms of the GNU General Public License as published by
##	the Free Software Foundation, either version 3 of the License, or
##	(at your option) any later version.
##
##	Fonty Python is distributed in the hope that it will be useful,
##	but WITHOUT ANY WARRANTY; without even the implied warranty of
##	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##	GNU General Public License for more details.
##
##	You should have received a copy of the GNU General Public License
##	along with Fonty Python.  If not, see <http://www.gnu.org/licenses/>.


import pathcontrol 
import fpversion
import os

copyright = "Fonty Python Copyright (C) 2006, 2007, 2008, 2009 Donn.C.Ingle"
contact = "email: donn.ingle@gmail.com"
done = "Done."

## We use PathControl to get some info, BUT we DO NOT WANT TO MAKE
## the ~/.fontypython folder from this module.
## This module (strings) also gets used via setup.py
## and that has the result of making the folder with root ownership! Bad news.
## So, there is a new flag to manage this now:
_pc = pathcontrol.PathControl( makeFolder=False )

version = _("Fonty Python version %s") % fpversion.version

warranty = """This program is distributed in the
hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.  See the GNU General Public License for
more details."""

aboutText = "%s\n%s\nWritten on Gnu\Linux, using wxPython.\n\n%s" % (copyright, version, warranty)

options=_("""Options:
  -v --version  Show program's version number and exit.
  -h --help	 Show this help message and exit.
  -e --examples
				Show some %$@#$ examples!
  -i Pog --install=Pog
				Install the fonts in this Pog to your 
				fonts folder.
  -u Pog --uninstall=Pog
				Uninstall the fonts in this Pog.
  -l --list	 List the names of all the Pogs.
  -s num --size=num
				Set a new default point size.
  -n num --number=num
				Set a new default for how many fonts 
				to view at one go. Don't overdo this.
  -t "text" --text="text"
				Set a new default sample text.  
				"This is an ex-parrot!", say. 
				Be sure to use the quotes.
  -p Pog --purge=Pog
				Purge the Pog of font files that are no
				longer really there.
  -c folder --check=folder
				Check for bad fonts that crash Fonty.
				It will recurse through sub-folders.
				This will build a file:
				~/.fontypython/segfonts
				After using this tool you should be
				able to browse folders that crashed
				Fonty. (The reason it's not done
				by default is that it's very slow.)
				* NOTE: The fonts that crash Fonty
				are probably still perfectly good
				and can be used in other apps. It's
				simply a bug in the library we use to
				access fonts that chokes things. This
				will (hopefully) improve in the future.
  -a folder Pog
  --all folder Pog
				Puts all fonts in this folder into the Pog.
				If the Pog already exists, it will add
				only *new* fonts, this means fonts are not
				repeated in that Pog.
  -A folder Pog
  --all-recurse folder Pog
				Puts all fonts in this folder and *all*
				sub-folders into the Pog. Rest same as -a.

  -z Pog --zip=Pog
				All the fonts inside Pog will be zipped
				and the zipfile will be named after the Pog.
				The file will be placed in the current
				directory.
				""")

use = _("""%(c)s [OPTIONS] [VIEW] [TARGET]
VIEW   : A place where fonts are. (A Pog or a Folder.)
TARGET : A "Pog". A place to keep those fonts.

("%(c)s" on it's own will start the GUI.)

NB: Try not to use spaces in Pog names. If you must, 
then "quote the name."

Please use -e to see more info.

NEWS : We now support TTF, OTF, Type1 (PFB, PFA) 
and TTC fonts.

%(version)s

The basic idea:
===============
Many designers have collections of font files in big
directory structures or on other media. Fonty Python
will let you gather your fonts and structure them
into collections -- or what I call "Pogs" -- a place
to keep tyPOGraphy. Well, why not?

Your fonts never move from where they are
(so don't worry). All that happens is that you select 
fonts visually and place their names into a Pog,
then you install or uninstall Pogs as you need them.
No copies of your fonts are made, only links to the
original files are used to install the fonts.

For example, you might have a Pog called "logos"
into which you place all the fonts you have of
company logos. After that, when you need to work
with those logos, you simply install the 'logos' Pog
and start your design app!

FP is also great for just looking at fonts wherever
they are on your computer, without having to install
them system-wide.

Manage your fonts on Gnu/Linux!
===============================
%(copy)s

%(warranty)s

%(contact)s
""" ) % { "c":"fontypython", "folder":_pc.appPath(), "contact":contact, "copy":copyright, "warranty":warranty, "version":version }

	
examples = _("""The basic format is:
%(c)s [VIEW] [TARGET]
  VIEW   = A place where fonts are. A Pog or a folder
		   someplace.
  TARGET = A Pog, a place to keep references to fonts
		   If you don't include a target then you are
		   viewing/editing only.

Tips:
=====
* Don't use spaces in Pog names. If you absolutely
  must then use quotes around the name, e.g. "Pogs
  of Ni"
* If your design apps (for example The Gimp) do not
  reflect the fonts that you have installed, restart
  the app. Sometimes the system needs a while to
  reflect the new fonts in your fonts folder.

Examples: All using short options, see -h
=========
%(c)s /path/to/fonts/ttfs/a
  This will start off showing the fonts in that path.
 
%(c)s /path/to/fonts/ttfs/b Trouser
  This will let you view and choose fonts from the
  path and it will store them in a Pog named "Trouser".
  The Pog will be created if it's not already there.

%(c)s Lumberjack
  This will let you see the fonts in the Pog named
  "Lumberjack". You can also uninstall individual
  fonts by selecting them. A cross will appear
  indicating the fonts that will be uninstalled.

%(c)s Camelot Spamalot
  This will let you see and choose fonts in
  "Camelot" and it will store them in "Spamalot"
  It lets you copy fonts between Pogs.

%(c)s -i Cheese
  Will install the fonts in Pog Cheese so you can
  use them in other apps.

%(c)s -u Trouser
  Will uninstall the fonts listed in Pog Trouser,
  so you can't use 'em anymore.( You Naughty thing) 

%(c)s -t "Pigs on the wing"
  Will set the text and exit. It's odd that way.
  Restart Fonty again to see the change.
  * You can also do this from the gui, without a
  restart.
  
%(c)s -s 128 
  Will set the point size to 128 - Crazy man!

%(c)s -v 25
  Will show 25 fonts at a time. Beware large numbers!
	
%(c)s -s 64 -v 10 Pimple
  Will set the point size to 64, the number of fonts
  to view is 10 and then display the Pimple Pog.

%(c)s -p Glutton
  If there are any fonts in "Glutton" that are not
  really on your drive/media anymore (perhaps you
  deleted them or the cat did) then this will go 
  through your Pog and cull them.

%(c)s -c /some/path/to/fonts
  If Fonty keeps crashing on /some/path/to/fonts
  then you should run a check on that folder.
  This will 'mark' the dangerous fonts and let
  you use that folder in the future.

%(c)s -a /some/path HolyHandGrenade
  This will put all the fonts in that path into
  the Pog called HolyHandGrenade.

%(c)s -A /some/path Tutto
  This will do the same as -a above: start in that 
  path, but it will then walk down recursivly 
  through all sub-folders too. The fonts will 
  be placed in Tutto.

Your fontypython folder is:
%(folder)s
If you want to backup your Pogs, that's where ya go.
%(contact)s

%(copy)s""") % { "c":"fontypython", "folder":_pc.appPath(), "contact":contact, "copy":copyright }

## These two are used in setup.py
description = _("Fonty Python - view and manage all kinds of fonts on Gnu/Linux")
long_description = _("""Manage your fonts on Gnu/Linux.
NEWS : We now support TTF, OTF, Type1 (PFB, PFA) 
and TTC fonts.

Many designers have collections of font files in big
directory structures or on other media. Fonty Python
will let you gather your fonts and structure them
into collections -- or what I call "Pogs" -- a place
to keep tyPOGraphy. Well, why not?

Your fonts never move from where they are
(so don't worry). All that happens is that you select 
fonts visually and place their names into a Pog,
then you install or uninstall Pogs as you need them.
No copies of your fonts are made, only links to the
original files are used to install the fonts.

For example, you might have a Pog called "logos"
into which you place all the fonts you have of
company logos. After that, when you need to work
with those logos, you simply install the 'logos' Pog
and start your design app!

FP is also great for just looking at fonts wherever
they are on your computer, without having to install
them system-wide.
	
	%(copy)s
	%(contact)s
	""") % {"copy":copyright, "contact":contact}


wxvers="2.8"
wxVersionError = _("""I cannot find "python-wxversion"
Please install this package - NB: ensure that
you use only the "Unicode build".

TIP
===
On my distro I can search for it like this:
aptitude search python-wx
This returns many results, one of which is:
python-wxversion 
I then install it like this:
sudo aptitude install python-wxversion 

If you get long error messages, you will need to
install python-wxgtk*, where the star means the 
version number and it should be at least %(wxv)s

You can also get the latest version from here:
http://wxpython.org/download.php
""") % {"wxv":wxvers}
	
wxError =_("""I cannot find "python-wxgtkX.Y"
Please install this package - NB: ensure that
you use only the "Unicode build".

TIP
===
On my distro I can search for it like this:
aptitude search python-wx
This returns many results, one of which is:
python-wxgtk%(wxv)s
I then install it like this:
sudo aptitude install python-wxgtk%(wxv)s

Make sure it's at least version %(wxv)s

You can also get the latest version from here:
http://wxpython.org/download.php
""") % {"wxv":wxvers}


PILError = _("""I cannot find "python-imaging"
Please install this package.

TIP
===
On my distro I can search for it like this:
aptitude search python-imag
This returns many results, one of which is:
python-imaging
I then install it like this:
sudo aptitude install python-imaging

Make sure it's at least version 1.1.6-1

You can also get the latest version from here:
http://www.pythonware.com/products/pil/index.htm
""")

## Won't xlate the thanks:
thanks = u"""Many thanks to:
1. Robin Dunn - wxPython and much sundry help.

2. Martin v. Löwis - Essential concepts regarding unicode and files.

3. Pietro Battiston - Italian translation.

4. Baptiste - French translation and many ideas.

5. Jason Yamada-Hanff - For the wiki, at least :)

6. Michael Hoeft - For code, friendship and the German translation.

7. Kartik Mistry - Our esteemed Debian packager!

8. savannah.nongnu.org - For the hosting.

9. And all those I have neglected to include.
"""

## June 2009 : Get the GPL from the COPYING file rather than a copy of it all here again.
try:
	root = __file__                              
	if os.path.islink(root):                     
		root = os.path.realpath(root)            
	fontyroot = os.path.dirname(os.path.abspath(root)) 
	p = os.path.join(fontyroot,'COPYING')
	GPL = open(p,"r").read()
except:
	GPL = """
					GNU GENERAL PUBLIC LICENSE
					   Version 3, 29 June 2007

The file "COPYING" cannot be found. Please check the installation directory for the licence.
"""
